### PR TITLE
Use Custom SSZ for P2P Types

### LINKED ISSUE #7425: Req-resp Error responses are decoded incorrectly.
# 🐞 Bug Report

### Description

`ErrorResponse` is defined in the spec as a `List[byte, 256]`, not a `Container{ List[byte, 256] }`. The spec states:

> ssz_snappy: The contents are first SSZ-encoded and then compressed with Snappy frames compression. For objects containing a single field, only the field is SSZ-encoded not a container with a single field. For example, the BeaconBlocksByRoot request is an SSZ-encoded list of Root's. This encoding type MUST be supported by all clients.

And defines the error as:

> The ErrorMessage schema is:
>
>   (
>     error_message: List[byte, 256]
>   )
>
> Note: By convention, the error_message is a sequence of bytes that MAY be interpreted as a UTF-8 string (for debugging purposes). Clients MUST treat as valid any byte sequences.

The schema used in Prysm, and the type with generated FastSSZ methods, incorrectly deserialize as container.

The difference is that the first 4 bytes are being recognized as offset in the container case. And for such single-field container, it would always be `0x00000004`. Hence the spec actively avoids adding useless wrappers on responses like that.

This results in all standard-format req-resp errors not being readable when decoded in Prysm (since Prysm expects that magic useless offset), as well as all produced errors starting with zeroes when read on another client (and potentially being completely invalid because of the added 4 bytes, i.e. 260 bytes total).

### Has this worked before in a previous version?

Not that I know.

## 🔬 Minimal Reproduction

-

## 🔥 Error

None.

## 🌍  Your Environment

**What version of Prysm are you running? (Which release)**

See relevant code and commit here: https://github.com/prysmaticlabs/prysm/blob/70f3fcdbd9e861dab869c7a2062f989b5ceca222/proto/beacon/p2p/v1/generated.ssz.go#L1475

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature Addition And Bug Fix

**What does this PR do? Why is it needed?**

- [x] This adds ssz support for non composite types. Previously we used `go-ssz` for non composite type which 
led our node to be susceptible to any go-ssz specific bugs. This PR instead implements custom ssz for our 
required p2p types, so that they do not need to use go-ssz anymore
- [x] Add in relevant tests.
- [x] Fix usage of this across the repo.

**Which issues(s) does this PR fix?**

Fixes #7425 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/p2p/types/types.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	return append(dst, marshalledObj...), nil
+}
+
+// MarshalSSZ Marshals the block by roots request type into the serialized object.
+func (s *BeaconBlockByRootsReq) MarshalSSZ() ([]byte, error) {
+	if len(*s) > int(params.BeaconNetworkConfig().MaxRequestBlocks) {
+		return nil, errors.Errorf("beacon block by roots request exceeds max size: %d > %d", len(*s), params.BeaconNetworkConfig().MaxRequestBlocks)
+	}
+	buf := make([]byte, 0, s.SizeSSZ())
+	for _, r := range *s {
```
Comment (prestonvanloon) [inline beacon-chain/p2p/types/types.go:67]: Do you have to deference the pointer of s?

File: beacon-chain/sync/error_test.go [hunk 2]
```diff
@@ -4,8 +4,9 @@ import (
 	"bytes"
 	"testing"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p/types"
+
```
Comment (prestonvanloon) [inline beacon-chain/sync/error_test.go:8]:
```suggestion
```

File: beacon-chain/p2p/types/types.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/p2p/types/types.go:67]: You cant loop through a pointer of a slice, which is why it needs to be dereferenced here.

File: proto/beacon/p2p/v1/BUILD.bazel [hunk 3]
```diff
@@ -26,8 +26,6 @@ ssz_gen_marshal(
     ],
     objs = [
         "BeaconBlocksByRangeRequest",
-        "BeaconBlocksByRootRequest",
```
Comment (terencechain) [inline proto/beacon/p2p/v1/BUILD.bazel:29]: Delete `BeaconBlocksByRootRequest` in `messages.proto`?

File: beacon-chain/p2p/types/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,25 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+load("@prysm//tools/go:def.bzl", "go_library")
+
+go_library(
+    name = "go_default_library",
+    srcs = ["types.go"],
+    importpath = "github.com/prysmaticlabs/prysm/beacon-chain/p2p/types",
+    visibility = ["//visibility:public"],
```
Comment (terencechain) [inline beacon-chain/p2p/types/BUILD.bazel:8]: Should the visibility be public here?

Comment (nisdas) [reply beacon-chain/p2p/types/BUILD.bazel]: nice catch ! that is why the CI was complaining lol

File: proto/beacon/p2p/v1/BUILD.bazel (hunk 3 above)
Comment (nisdas) [reply proto/beacon/p2p/v1/BUILD.bazel:29]: Its been deleted form before, this is just cleanup

File: beacon-chain/p2p/types/types.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/p2p/types/types.go:67]: Makes sense thanks

**Review by prestonvanloon — APPROVED** [review]

**Comment by prestonvanloon** [comment]
Conflicts :sob: PTAL @nisdas

**Review by prestonvanloon — APPROVED** [review]

**Review by prestonvanloon — DISMISSED** [review]

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**